cmd/xl-v1-healing.go

```diff
@@ -82,29 +86,56 @@ func (xl xlObjects) HealBucket(bucket string) error {
 	}); reducedErr != nil {
 		return toObjectErr(reducedErr, bucket)
 	}
-	return nil
+
+	// Proceed to heal bucket metadata.
+	return healBucketMetadata(xl.storageDisks, bucket)
 }
 
-// HealObject heals a given object for all its missing entries.
-// FIXME: If an object object was deleted and one disk was down, and later the disk comes back
-// up again, heal on the object should delete it.
-func (xl xlObjects) HealObject(bucket, object string) error {
-	// Verify if bucket is valid.
-	if !IsValidBucketName(bucket) {
-		return traceError(BucketNameInvalid{Bucket: bucket})
+// Heals all the metadata associated for a given bucket.
+func healBucketMetadata(storageDisks []StorageAPI, bucket string) error {
+	policyPath := pathJoin(bucketConfigPrefix, bucket, policyJSON)
+	// Lock the `policy.json` before healing.
+	policyLock := nsMutex.NewNSLock(minioMetaBucket, policyPath)
+	policyLock.RLock()
+	defer policyLock.RUnlock()
+
```

Comment:
> The reason to not hold Write lock is that it would block the Read operations, also since we are only writing to missing disks. Additional advantage here is that Write() outside of this call at this path will still block because of our Read lock. So all in all it works without much trouble. 

Ok. The code is however holding onto the lock a bit longer than necessary - e.g. policy path lock is held during notification config heal, and notification path lock is held during listener config heal.